src/components/flame-graph/Canvas.js

```diff
@@ -48,6 +52,64 @@ const ROW_HEIGHT = 16;
 const TEXT_OFFSET_START = 3;
 const TEXT_OFFSET_TOP = 11;
 
+/* Return the background and foreground colors used to paint a box in
+ * the flame graph. The passed stack type determines the hue while the
+ * self time determines the lightness of the color. Darker colors
+ * indicate higher self times. */
+export function getColors(
+  stackType: StackType,
+  selfTimeRelative: number
+): {| background: string, foreground: string |} {
+  /* The background colors below are defined as a range where the
+   * starting color is the color shown when the function has no self
+   * time at all, and the ending color is when all available self time
+   * (the total time) in the flame graph is assigned to a single
+   * function. This latter case occurs when there's only a single box
+   * at the top, stretching over all horizontal space in the graph,
+   * covering all boxes below it. When the self time lies in between
+   * these two extremes, the color is an interpolated value lying
+   * between the starting and ending colors in the range.
+   *
+   * The interpolated color is derived from `selfTimeRelative`, which
+   * is the self time of the frame relative to all available self time
+   * possibly represented in the flame graph. However, using this
+   * value to interpolate to a color linearly will require the value
+   * to be relatively big in order to be able to see the shift in
+   * color, and such large self times would be visually apparent from
+   * the shape of the flame graph anyway. We are better off if we can
+   * discern also small values of self time from the color, and this
+   * is accomplished by changing the linear interpolation to a
+   * logarithmic one instead, with a steep increase in the beginning.
+   *
+   * Below we use the transformation y(x) = log(c * x + 1) / log(c + 1),
+   * where c is a constant. It has the property that y(0) = 0 and
+   * y(1) = 1. The constant has been chosen by visual inspection of
+   * the graph. */
+
+  let backgroundRange: [string, string];
+  let foreground: string;
+
+  const t = Math.log(5000 * selfTimeRelative + 1) / Math.log(5001);
+
+  switch (stackType) {
+    case 'native':
+      backgroundRange = [colors.BLUE_40, colors.BLUE_70];
+      foreground = t > 0.25 ? '#ffffff' : colors.BLUE_90;
+      break;
+    case 'js':
+      backgroundRange = [colors.ORANGE_50, colors.ORANGE_70];
+      foreground = t > 0.33 ? '#ffffff' : colors.ORANGE_90;
+      break;
+    case 'unsymbolicated':
+      backgroundRange = [colors.GREY_30, colors.GREY_50];
+      foreground = t > 0.33 ? '#ffffff' : colors.GREY_90;
+      break;
+    default:
+      throw new Error(`Unknown stack type case "${(stackType: empty)}".`);
+  }
+  return { background: interpolateRgb(...backgroundRange)(t), foreground };
```

Comment:
D3 also has `interpolateRgbBasis` (which allows interpolation between several points, like a list of `[ORANGE_50, ORANGE_60, ORANGE_70]`) and `interpolateHsl`, but `interpolateRgb` seemed to be (after making a few informal measurements) the fastest, and provided more or less the same color ranges.

By the way, `getColors` seems to be responsible of 20% of the time in `_drawCanvas`. Not optimal, but perhaps alright for a first version? Most of it seemingly having to do with converting colors to RGB within in D3.

Reply:
Oh, and one other comment, if you can split out the foreground and background colors into separate functions, that'll save returning an object. I'm not sure the actual impact of JITed code returning objects that immediately get destructured, it may optimize that allocation out (no idea really.) I think this type of thing is important due to the perf-sensitive nature of the function.